File: example/python/product/e2.py
import argparse
import airbot_hardware_py as ah
import time
import threading
import json
import socket
import math

# UDP server configuration (PlotJuggler's listening address)
UDP_SERVER_IP = "127.0.0.1"
UDP_SERVER_PORT = 9870

# Create a UDP socket
sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
server_address = (UDP_SERVER_IP, UDP_SERVER_PORT)

print(f"Sending data to {UDP_SERVER_IP}:{UDP_SERVER_PORT}")


def send_to_plotjuggler(eef, stop_event):
    while not stop_event.is_set():
        current_time = time.time()
        state_positions = 0.0

        if eef.state().is_valid:
            state_positions = eef.state().pos

        # Create the JSON object
        json_obj = {
            "state/e2_position": state_positions,
            "timestamp": current_time,
        }

        # Serialize the JSON object to a string and then encode to bytes
        message = json.dumps(json_obj).encode("utf-8")

        try:
            sock.sendto(message, server_address)
        except Exception as e:
            print(f"Failed to send state data: {e}")

        # Wait before the next round of prints
        time.sleep(0.01)


def main():
    # 设置命令行参数解析
    parser = argparse.ArgumentParser(description="单独 E2 位置反馈")
    parser.add_argument(
        "-i", "--can", default="can0", help="can口名称 (如: can0, can1)"
    )

    args = parser.parse_args()

    # 获取解析后的参数值
    can_interface = args.can

    executor = ah.create_asio_executor(1)
    io_context = executor.get_io_context()

    eef = ah.EEF1.create(ah.EEFType.E2, ah.MotorType.OD)
    if not eef.init(io_context, can_interface, 250):
        raise RuntimeError("The E2 initialization failed.")

    eef.enable()
    # eef.set_param("eef.control_mode", ah.MotorControlMode.MIT)
    # eef.set_param("eef.e2.mode", ah.ParamType.UINT32_LE)
    # eef.set_param("eef.e2.mode", 4)
    param_value = ah.ParamValue(ah.ParamType.UINT32_LE, 4)
    eef.set_param("eef.e2.mode", param_value)

    stop_event = threading.Event()
    print_thread = threading.Thread(target=send_to_plotjuggler, args=(eef, stop_event))
    print_thread.daemon = True
    print_thread.start()

    try:
        while True:
            eef.mit(ah.EEFCommand1())
            if eef.state().is_valid:
                print(f"The parallel width of E2: {eef.state().pos[0]:.3f}")

            time.sleep(0.004)
    except KeyboardInterrupt:
        print("\n检测到Ctrl+C，正在停止机械臂并释放资源...")
    finally:
        stop_event.set()
        if print_thread.is_alive():
            print_thread.join()

        eef.disable()
        eef.uninit()


if __name__ == "__main__":
    main()

File: example/cpp/src/product/e2.cpp
#include <argparse/argparse.hpp>
#include <boost/asio.hpp>
#include <boost/json.hpp>
#include <chrono>
#include <cmath>
#include <iostream>
#include <thread>
#include <utility>

#include "airbot_hardware/executors/executor.hpp"
#include "airbot_hardware/handlers/eef.hpp"

using namespace std::chrono_literals;
using namespace airbot::hardware;

// Function to create state JSON
boost::json::object stateToJson(const std::vector<double>& positions, const std::vector<double>& velocity,
                                const std::vector<double>& effort) {
  boost::json::object json_obj;

  // Convert vectors to JSON arrays
  boost::json::array pos_array;
  boost::json::array vel_array;
  boost::json::array eff_array;
  for (const auto& pos : positions) pos_array.push_back(pos);
  for (const auto& vel : velocity) vel_array.push_back(vel);
  for (const auto& eff : effort) eff_array.push_back(eff);

  json_obj["position"] = pos_array;
  json_obj["velocity"] = vel_array;
  json_obj["effort"] = eff_array;
  json_obj["timestamp"] = static_cast<double>(std::chrono::duration_cast<std::chrono::microseconds>(
                                                  std::chrono::steady_clock::now().time_since_epoch())
                                                  .count()) /
                          1e6;

  return json_obj;
}

int main(int argc, char* argv[]) {
  argparse::ArgumentParser program("airbot_e2", "1.0");

  program.add_argument("-i", "--can-interface")
      .help("CAN interface name (e.g., can0, can1)")
      .default_value(std::string("can0"));

  program.add_argument("-r", "--remote-ip").help("Plotjuggler IP address").default_value(std::string("127.0.0.1"));

  program.add_argument("-p", "--port").help("UDP port for Plotjuggler").scan<'i', int>().default_value(9870);

  try {
    program.parse_args(argc, argv);
  } catch (const std::runtime_error& err) {
    std::cerr << err.what() << std::endl;
    std::cerr << program;
    return 1;
  }

  std::string can_interface = program.get<std::string>("--can-interface");
  std::string remote_ip = program.get<std::string>("--remote-ip");
  int port = program.get<int>("--port");

  std::cout << "Using CAN interface: " << can_interface << std::endl;
  std::cout << "Sending data to Plotjuggler at " << remote_ip << ":" << port << std::endl;

  auto executor = AsioExecutor::create(1);
  auto eef = EEF<1>::create<EEFType::E2, MotorType::OD>();

  eef->init(executor->get_io_context(), can_interface.c_str(), 250_hz);
  eef->enable();
  eef->set_param("control_mode", static_cast<uint32_t>(MotorControlMode::MIT));

  boost::asio::io_context io_context;
  boost::asio::ip::udp::socket socket(io_context);
  boost::asio::ip::udp::endpoint server_endpoint(boost::asio::ip::make_address(remote_ip), port);

  try {
    socket.open(boost::asio::ip::udp::v4());
  } catch (std::exception& e) {
    std::cerr << "Failed to open UDP socket: " << e.what() << std::endl;
    eef->disable();
    eef->uninit();
    return 1;
  }

  auto ts = std::thread([&]() {
    auto last_print_time = std::chrono::steady_clock::now();

    while (true) {
      auto now = std::chrono::steady_clock::now();

      if (now - last_print_time > 100ms) {
        auto state = eef->state();
        double width = state.pos[0];
        std::cout << "E2 Parallel Width: " << width << std::endl;
        last_print_time = now;
      }

      auto state = eef->state();
      std::vector<double> pos(state.pos.begin(), state.pos.end());
      std::vector<double> vel(state.vel.begin(), state.vel.end());
      std::vector<double> eff(state.eff.begin(), state.eff.end());

      std::string state_string = boost::json::serialize(stateToJson(pos, vel, eff));

      try {
        socket.send_to(boost::asio::buffer(state_string), server_endpoint);
      } catch (std::exception& e) {
        std::cerr << "Failed to send state data: " << e.what() << std::endl;
      }

      std::this_thread::sleep_for(10ms);
    }
  });

  ts.detach();

  auto start = std::chrono::steady_clock::now();
  bool running = true;

  while (running && std::chrono::steady_clock::now() - start < 120s) {
    eef->mit({0.0, 0.0, 0.0, 0.0, 0.0, 0.0});
    std::this_thread::sleep_for(1ms);
  }

  eef->disable();
  eef->uninit();

  return 0;
}
